kolibri/auth/test/test_permissions.py

```diff
@@ -1,454 +1,568 @@
-#         self.assertFalse(self.coach1.has_perm(self.AUTH_REMOVE_LEARNER, self.learner_groups[1]))
-
-#     def test_remove_learner_specific_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_LEARNER, self.learner_groups[0]))
-
-#     def test_remove_learner_rejects_non_classroom_objects(self):
-#         with self.assertRaises(InvalidPermission):
-#             self.admin.has_perm(self.AUTH_REMOVE_LEARNER, obj={})
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.add_facility_admin     ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_ADD_FACILITY_ADMIN = 'auth.add_facility_admin'
-
-#     def test_add_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     def test_add_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     def test_add_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_ADD_FACILITY_ADMIN))
-
-#     # noqa ##################################
-#     # noqa #                               ##
-#     # noqa #   auth.remove_facility_admin  ##
-#     # noqa #                               ##
-#     # noqa ##################################
-#     AUTH_REMOVE_FACILITY_ADMIN = 'auth.remove_facility_admin'
-
-#     def test_remove_facility_admin_for_admin(self):
-#         self.assertTrue(self.admin.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_coach(self):
-#         self.assertFalse(self.coach2.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
-
-#     def test_remove_facility_admin_for_learner(self):
-#         self.assertFalse(self.learner1.has_perm(self.AUTH_REMOVE_FACILITY_ADMIN))
+"""
+Tests of the permissions on specific models in the auth app. For tests of the permissions system itself, see test_permission_classes.py
+"""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+from django.test import TestCase
+
+from .helpers import create_dummy_facility_data
+
+from ..constants import role_kinds
+from ..models import DeviceOwner, Facility, Classroom, LearnerGroup, Role, FacilityUser
+
+
+class FacilityPermissionsTestCase(TestCase):
+    """
+    Tests of permissions for reading/modifying Facility instances
+    """
+
+    def setUp(self):
+        self.data1 = create_dummy_facility_data()
+        self.data2 = create_dummy_facility_data()
+        self.device_owner = DeviceOwner.objects.create(username="boss")
+
+    def test_facility_users_cannot_create_facility(self):
+        """ FacilityUsers can't create new Facilities, regardless of their roles """
+        unsaved_facility = Facility(name="Home")
+        self.assertFalse(unsaved_facility.user_can_create(self.data1["facilityadmin"]))
```

Comment:
I hear what you're saying, but here's my take:

- I believe it is standard practice (e.g. in Django Rest Framework) when checking row-level permissions, to check them against an instantiated model instance.
- I don't see a way around it, given that many permission checks rely on running methods or reading attributes on the model class.
- The same situation exists for reading (and updating, deleting, etc) -- there's nothing stopping a developer from doing stuff (e.g. reading some data, and doing something based on it) prior to checking the permissions. Care needs to be taken, and tests (e.g. of a view) can be written to check for violations, where appropriate.
- The majority of data access will happen through the HTTP API, from our frontend apps. In these cases, the developer will not be responsible for doing any checking, as it will be handled seamlessly by Django REST Framework.